File: src/lib/api/knowledge.ts
import { apiFetch } from "./client";

export type SourceType = "note" | "mistake";

export type RelationType =
  | "explains"
  | "similar"
  | "prerequisite"
  | "follow_up"
  | "source_for";

export type RelationStatus = "suggested";

export type ReviewStateFilter = "due" | "overdue" | "upcoming" | "all";

export type CitationBlockType =
  | "source_backed_claim"
  | "ai_inference"
  | "insufficient_context";

export interface SourceRef {
  source_type: SourceType;
  source_id: string;
  title?: string;
  slug?: string;
  field?: string;
  excerpt?: string;
  url?: string;
  confidence?: number;
  match_reasons?: string[];
}

export interface RelationSuggestion {
  source_type: SourceType;
  source_id: string;
  target_type: SourceType;
  target_id: string;
  relation_type: RelationType;
  score: number;
  reason: string;
  status: RelationStatus;
}

export interface NoteBrief {
  id: string;
  slug: string;
  title: string;
  type: string;
  subject?: string | null;
  knowledge_points?: string | null;
  difficulty?: string | null;
  summary?: string | null;
  updated_at?: string | null;
}

export interface ContextPackStats {
  mistake_count: number;
  note_count: number;
  top_error_reasons: string[];
}

export interface ContextPackRequest {
  subject?: string;
  knowledge_points?: string[];
  tags?: string[];
  type?: string;
  difficulty?: string;
  date_range?: { from?: string; to?: string };
  review_state?: ReviewStateFilter;
  limit?: number;
}

export interface ContextPackResponse {
  sources: SourceRef[];
  related_notes: NoteBrief[];
  related_mistakes: NoteBrief[];
  suggested_relations: RelationSuggestion[];
  stats: ContextPackStats;
}

export interface WeakPointItem {
  subject: string;
  knowledge_point: string;
  mistake_count: number;
  due_review_count: number;
  recent_error_count: number;
  top_error_reasons: string[];
  evidence_sources: SourceRef[];
}

export interface WeakPointsResponse {
  days: number;
  weak_points: WeakPointItem[];
}

export interface CitationBlock {
  type: CitationBlockType;
  text: string;
  source_refs: SourceRef[];
}

export interface KnowledgeSummaryRequest {
  mode?: string;
  context_pack: ContextPackResponse;
  requirements?: {
    language?: string;
    style?: string;
    max_length?: number;
    require_citations?: boolean;
  };
}

export interface KnowledgeSummaryResponse {
  title: string;
  blocks: CitationBlock[];
}

export interface InsufficientContextResponse {
  status: "insufficient_context";
  message: string;
  outline: string[];
}

export async function getContextPack(
  request: ContextPackRequest
): Promise<ContextPackResponse> {
  return apiFetch<ContextPackResponse>("/api/knowledge/context-pack", {
    method: "POST",
    body: JSON.stringify(request),
  });
}

export async function getWeakPoints(
  days: number = 30
): Promise<WeakPointsResponse> {
  return apiFetch<WeakPointsResponse>(
    `/api/knowledge/weak-points?days=${days}`
  );
}

export async function getKnowledgeSummary(
  request: KnowledgeSummaryRequest
): Promise<KnowledgeSummaryResponse | InsufficientContextResponse> {
  return apiFetch<KnowledgeSummaryResponse | InsufficientContextResponse>(
    "/api/ai/knowledge-summary",
    {
      method: "POST",
      body: JSON.stringify(request),
    }
  );
}

File: backend/app/schemas/knowledge.py
import uuid
from datetime import date, datetime
from enum import Enum

from pydantic import BaseModel, Field


class SourceType(str, Enum):
    note = "note"
    mistake = "mistake"


class RelationType(str, Enum):
    explains = "explains"
    similar = "similar"
    prerequisite = "prerequisite"
    follow_up = "follow_up"
    source_for = "source_for"


class RelationStatus(str, Enum):
    suggested = "suggested"


class ReviewStateFilter(str, Enum):
    due = "due"
    overdue = "overdue"
    upcoming = "upcoming"
    all = "all"


class SourceRef(BaseModel):
    source_type: SourceType
    source_id: str
    title: str = ""
    slug: str = ""
    field: str = ""
    excerpt: str = ""
    url: str = ""
    confidence: float = 0.0
    match_reasons: list[str] = Field(default_factory=list)


class RelationSuggestion(BaseModel):
    source_type: SourceType
    source_id: str
    target_type: SourceType
    target_id: str
    relation_type: RelationType
    score: float = 0.0
    reason: str = ""
    status: RelationStatus = RelationStatus.suggested


class DateRange(BaseModel):
    from_date: date | None = Field(None, alias="from")
    to_date: date | None = Field(None, alias="to")

    model_config = {"populate_by_name": True}


class ContextPackRequest(BaseModel):
    subject: str | None = None
    knowledge_points: list[str] = Field(default_factory=list)
    tags: list[str] = Field(default_factory=list)
    type: str | None = None
    difficulty: str | None = None
    date_range: DateRange | None = None
    review_state: ReviewStateFilter | None = None
    limit: int = Field(default=10, ge=1, le=100)


class ContextPackStats(BaseModel):
    mistake_count: int = 0
    note_count: int = 0
    top_error_reasons: list[str] = Field(default_factory=list)


class NoteBrief(BaseModel):
    id: uuid.UUID
    slug: str
    title: str
    type: str
    subject: str | None = None
    knowledge_points: str | None = None
    difficulty: str | None = None
    summary: str | None = None
    updated_at: datetime | None = None

    model_config = {"from_attributes": True}


class ContextPackResponse(BaseModel):
    sources: list[SourceRef] = Field(default_factory=list)
    related_notes: list[NoteBrief] = Field(default_factory=list)
    related_mistakes: list[NoteBrief] = Field(default_factory=list)
    suggested_relations: list[RelationSuggestion] = Field(default_factory=list)
    stats: ContextPackStats = Field(default_factory=ContextPackStats)


class WeakPointItem(BaseModel):
    subject: str
    knowledge_point: str
    mistake_count: int = 0
    due_review_count: int = 0
    recent_error_count: int = 0
    top_error_reasons: list[str] = Field(default_factory=list)
    evidence_sources: list[SourceRef] = Field(default_factory=list)


class WeakPointsResponse(BaseModel):
    days: int
    weak_points: list[WeakPointItem] = Field(default_factory=list)


class CitationBlockType(str, Enum):
    source_backed_claim = "source_backed_claim"
    ai_inference = "ai_inference"
    insufficient_context = "insufficient_context"


class CitationBlock(BaseModel):
    type: CitationBlockType
    text: str = ""
    source_refs: list[SourceRef] = Field(default_factory=list)


class KnowledgeSummaryRequest(BaseModel):
    mode: str = "exam_review_summary"
    context_pack: ContextPackResponse
    requirements: dict = Field(default_factory=lambda: {
        "language": "zh-CN",
        "style": "exam_review",
        "max_length": 1200,
        "require_citations": True,
    })


class KnowledgeSummaryResponse(BaseModel):
    title: str = ""
    blocks: list[CitationBlock] = Field(default_factory=list)


class InsufficientContextResponse(BaseModel):
    status: str = "insufficient_context"
    message: str = "No usable source references are available for factual generation."
    outline: list[str] = Field(default_factory=list)
